src/presentation/service/services.go

```diff
@@ -0,0 +1,470 @@
+	servicesInfra "github.com/speedianet/os/src/infra/services"
+	vhostInfra "github.com/speedianet/os/src/infra/vhost"
+	mappingInfra "github.com/speedianet/os/src/infra/vhost/mapping"
+	serviceHelper "github.com/speedianet/os/src/presentation/service/helper"
+)
+
+type ServicesService struct {
+	persistentDbService *internalDbInfra.PersistentDatabaseService
+}
+
+func NewServicesService(
+	persistentDbService *internalDbInfra.PersistentDatabaseService,
+) *ServicesService {
+	return &ServicesService{
+		persistentDbService: persistentDbService,
+	}
+}
+
+func (service *ServicesService) Read() ServiceOutput {
+	servicesQueryRepo := servicesInfra.NewServicesQueryRepo(service.persistentDbService)
+	servicesList, err := useCase.ReadServicesWithMetrics(servicesQueryRepo)
+	if err != nil {
+		return NewServiceOutput(InfraError, err.Error())
+	}
+
+	return NewServiceOutput(Success, servicesList)
+}
+
+func (service *ServicesService) ReadInstallables() ServiceOutput {
+	servicesQueryRepo := servicesInfra.NewServicesQueryRepo(service.persistentDbService)
+	servicesList, err := useCase.ReadInstallableServices(servicesQueryRepo)
+	if err != nil {
+		return NewServiceOutput(InfraError, err.Error())
+	}
+
+	return NewServiceOutput(Success, servicesList)
+}
+
+func (service *ServicesService) CreateInstallable(
+	input map[string]interface{},
+	shouldSchedule bool,
+) ServiceOutput {
+	requiredParams := []string{"name"}
+	err := serviceHelper.RequiredParamsInspector(input, requiredParams)
+	if err != nil {
+		return NewServiceOutput(UserError, err.Error())
+	}
+
+	name, err := valueObject.NewServiceName(input["name"])
+	if err != nil {
+		return NewServiceOutput(UserError, err.Error())
+	}
+
+	var versionPtr *valueObject.ServiceVersion
+	if input["version"] != nil {
+		version, err := valueObject.NewServiceVersion(input["version"])
+		if err != nil {
+			return NewServiceOutput(UserError, err.Error())
+		}
+		versionPtr = &version
+	}
+
+	var startupFilePtr *valueObject.UnixFilePath
+	if input["startupFile"] != nil {
+		startupFile, err := valueObject.NewUnixFilePath(input["startupFile"])
+		if err != nil {
+			return NewServiceOutput(UserError, err.Error())
+		}
+		startupFilePtr = &startupFile
+	}
+
+	envs := []valueObject.ServiceEnv{}
+	if input["envs"] != nil {
+		rawEnvs, assertOk := input["envs"].([]string)
+		if !assertOk {
+			return NewServiceOutput(UserError, "InvalidEnvs")
+		}
+
+		for _, rawEnv := range rawEnvs {
+			env, err := valueObject.NewServiceEnv(rawEnv)
+			if err != nil {
+				continue
+			}
+			envs = append(envs, env)
+		}
+	}
+
+	portBindings := []valueObject.PortBinding{}
+	if input["portBindings"] != nil {
+		rawPortBindings, assertOk := input["portBindings"].([]string)
+		if !assertOk {
+			return NewServiceOutput(UserError, "InvalidPortBindings")
+		}
+
+		for _, rawPortBinding := range rawPortBindings {
+			portBinding, err := valueObject.NewPortBinding(rawPortBinding)
+			if err != nil {
+				continue
```

Comment:
Here as well. Actually, in every place I discard something during a loop.

Reply:
Done.

Reply:
Wasn't done, there other areas on the code you forgot.

Reply:
Now it's done!